### PR TITLE
Miscellaneous bug fixes discovered by the sqllogictests.

### PR DESCRIPTION
**Pull request by petermattis**
I still have at least one more bug present in the "delete" tests.

See #2749.

### CODE REVIEW

File: sql/analyze.go [hunk 1]
```diff
@@ -195,7 +201,9 @@ outer:
 			if exprs[i] == nil {
 				// We found a simplification. Strip off the expression that is now nil
 				// and continue the outer loop.
-				exprs = exprs[:len(exprs)-1]
+				n := len(exprs) - 1
+				exprs[i], exprs[n] = exprs[n], exprs[i]
```
Comment (vivekmenezes) [inline sql/analyze.go]: exprs[i], exprs[n] = exprs[n], nil

File: sql/select.go [hunk 2]
```diff
@@ -405,6 +411,22 @@ func (v *indexInfo) makeConstraints(exprs []parser.Exprs) {
 						constraint.start = c
 					}
 				case parser.In:
+					// Only allow the IN constraint if the previous constraints are all
+					// EQ. This is necessary to prevent overlapping spans from being
+					// generated. Consider the constraints [a >= 1, a <= 2, b IN (1,
+					// 2)]. This would turn into the spans /1/1-/3/2 and /1/2-/3/3.
+					ok := true
+					for _, c := range v.constraints {
+						if c.start == c.end && c.start.Operator == parser.EQ {
```
Comment (vivekmenezes) [inline sql/select.go]: how about using the ! of this boolean expression and setting ok to false and breaking. Or set ok to the expression and break once ok is false.

**Comment by vivekmenezes** [comment]
LGTM

**Comment by tamird** [comment]
LGTM with comments (on commits) addressed

File: sql/analyze.go (hunk 1 above)
Comment (petermattis) [inline sql/analyze.go]: Actually, `exprs[i] = exprs[n]` works too and is even simpler.

File: sql/select.go (hunk 2 above)
Comment (petermattis) [inline sql/select.go]: I took the second option: `ok := c.start == c.end && c.Operator == parser.EQ; if !ok { break }`

**Pull request merged**